janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/LazyHasStep.java

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop.optimize;
+
+import org.apache.tinkerpop.gremlin.process.traversal.P;
+import org.apache.tinkerpop.gremlin.process.traversal.Traversal;
+import org.apache.tinkerpop.gremlin.process.traversal.step.TraversalParent;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.HasContainer;
+import org.apache.tinkerpop.gremlin.structure.Element;
+import org.apache.tinkerpop.gremlin.structure.util.StringFactory;
+
+import static org.janusgraph.graphdb.types.system.ImplicitKey.ADJACENT_ID;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class LazyHasStep<S extends Element, E> extends HasStep<S> implements TraversalParent {
+
+    private final Traversal.Admin<? extends Element, E> filterTraversal;
+
+    public LazyHasStep(Traversal.Admin traversal,
+                       final Traversal<? extends Element, E> filterTraversal) {
+        super(traversal);
+        this.filterTraversal = filterTraversal.asAdmin();
+        integrateChild(this.filterTraversal);
+
+        HasContainer hc = new LazyHasContainer(ADJACENT_ID.name(), this::initializeFilter);
+        this.addHasContainer(hc);
+    }
+
+    private P<?> initializeFilter() {
+        if (filterTraversal.hasNext()) {
+            return P.eq(filterTraversal.next());
```

Comment:
At this point, I'm not entirely sure if the `filterTraversal` uses the same transaction as the surrounding `traversal`. Does anyone know off the top of their heads, how such an access is handled by JanusGraph? And if not, can you hint me, where I have to look for it? So far, I was not able to find out, whether the same transaction is reused.

Reply:
While verifying that both traversals use the same transaction, I realized that not only those two transactions are equal. The same transaction is actually used in the optimization phase. With this knowledge, there is no need to wait until runtime before evaluating the `filterTraversal`. The matching ID can already be retrieved at optimization time and embedded into the query.

That makes the optimization *much* easier and also drops the need for custom classes like `LazyHasStep` and `LazyHasContainer`.